Fixing nil crashes

woody-apple opened the pull request:
Reported nil crashes affecting customers, fixing patterns across Matter.framework.

woody-apple commented on .vscode/settings.json:
```diff
@@ -133,7 +133,8 @@
         "variant": "cpp",
-        "any": "cpp"
+        "any": "cpp",
+        "future": "cpp"
```
This is what happens if you open any new file and save it in VSCode, so doing it for folks here

woody-apple commented on src/darwin/Framework/CHIP/MTRUtilities.mm:
```diff
@@ -19,5 +19,5 @@
 BOOL MTREqualObjects(id<NSObject> _Nullable a, id<NSObject> _Nullable b)
 {
-    return (a == nil) ? (b == nil) : [a isEqual:b];
+    return (a == nil) ? (b == nil) : (b == nil ? NO : [a isEqual:b]);
```
This is allowing a check to isEqual: nil, which isn't correct.

woody-apple commented on src/darwin/Framework/CHIP/MTRUtilities.mm:
```diff
@@ -19,5 +19,16 @@
 BOOL MTREqualObjects(id<NSObject> _Nullable a, id<NSObject> _Nullable b)
 {
-    return (a == nil) ? (b == nil) : [a isEqual:b];
+    // Check if A is nil, and return based on B being nil (or not)
```
Making this readable, and fixing where we can compare to nil

woody-apple commented on src/darwin/Framework/Matter.xcodeproj/project.pbxproj:
```diff
@@ -109,7 +109,7 @@
 		3D3928D72BBCEA3D00CDEBB2 /* MTRAvailabilityTests.m in Sources */ = {isa = PBXBuildFile; fileRef = 3D3928D62BBCEA3D00CDEBB2 /* MTRAvailabilityTests.m */; settings = {COMPILER_FLAGS = "-UMTR_NO_AVAILABILITY -Wno-unguarded-availability-new"; }; };
 		3D4733AF2BDF1B80003DC19B /* MTRSetupPayloadTests.m in Sources */ = {isa = PBXBuildFile; fileRef = 3D4733AE2BDF1B80003DC19B /* MTRSetupPayloadTests.m */; };
-		3D4733B32BE2D1DA003DC19B /* MTRUtilities.h in Headers */ = {isa = PBXBuildFile; fileRef = 3D4733B22BE2D1CF003DC19B /* MTRUtilities.h */; };
+		3D4733B32BE2D1DA003DC19B /* MTRUtilities.h in Headers */ = {isa = PBXBuildFile; fileRef = 3D4733B22BE2D1CF003DC19B /* MTRUtilities.h */; settings = {ATTRIBUTES = (Public, ); }; };
```
Working on this, our imports are broken. Need to move to xcconfigs, but... hopefully not in this PR to fix the crash.

woody-apple commented on src/darwin/Framework/CHIP/MTRBaseDevice.mm:
```diff
@@ -14,8 +14,7 @@
  */
-#import <Matter/MTRClusterConstants.h>
-#import <Matter/MTRDefines.h>
+#import <Matter/Matter.h>
```
This is also not the correct import, this should be a precompiled header for the project, but this'll patch it over for now.

woody-apple commented on src/darwin/Framework/CHIP/MTRBaseDevice.mm:
```diff
@@ -35,27 +34,27 @@
-#include <system/SystemClock.h>
-
-#include <memory>
+#import "app/ConcreteAttributePath.h"
```
These should all be imports, I'll be moving these to xc-configs in a follow on PR.

woody-apple commented on src/darwin/Framework/CHIP/Matter.h:
```diff
@@ -68,5 +68,6 @@
 #import <Matter/MTRStorage.h>
 #import <Matter/MTRStructsObjc.h>
 #import <Matter/MTRThreadOperationalDataset.h>
+#import <Matter/MTRUtilities.h>
```
We were missing an import, but we were extern-ing it. Adding it. I'll fix this file and others with a mass xcconfig fix.

woody-apple commented on src/darwin/Framework/CHIP/MTRDevice.mm:
```diff
@@ -41,17 +41,17 @@
-#include "lib/core/DataModelTypes.h"
-#include <app/ConcreteAttributePath.h>
-#include <lib/support/FibonacciUtils.h>
+#import "lib/core/CHIPError.h"
```
These should all be imports, I'll be moving these to xc-configs in a follow on PR.

woody-apple commented on src/darwin/Framework/CHIP/MTRDevice.mm:
```diff
@@ -15,7 +15,7 @@
  */
 
-#import <Matter/MTRDefines.h>
+#import <Matter/Matter.h>
```
This is the wrong way to import things, I'll fix this in a follow-on PR with general xcconfig and PCH (precompiled headers). This is a quick general patch.

bzbarsky-apple commented on src/darwin/Framework/CHIP/MTRUtilities.mm:
```diff
@@ -19,5 +19,16 @@
 BOOL MTREqualObjects(id<NSObject> _Nullable a, id<NSObject> _Nullable b)
 {
-    return (a == nil) ? (b == nil) : [a isEqual:b];
+    // Check if A is nil, and return based on B being nil (or not)
```
So here's what confuses me.  The new logic here is equivalent to the old logic, so that's fine as far as it goes.  But why did we need to change this at all?  isEqual is documented as:

```
object

    The object to be compared to the receiver. May be nil, in which case this method returns NO.
```

so if `b` is nil and we reached the `[a isEqual:b]` stuff we would return NO as desired.  The new code is more explicit about it, I guess?

woody-apple replied:
Yes it is, and readable.

woody-apple replied:
The old code allowed you to pass nil to [a isEqual: b];

bzbarsky-apple reviewed (commented):
So the real fix here is making sure we don't pass nil to `isEqualToNumber:`, right?  The rest is fixing other correctness issues and/or other issues altogether.

Would be good for the commit message to make this clear....

woody-apple commented on src/darwin/Framework/CHIP/MTRBaseDevice.mm:
```diff
@@ -35,27 +34,27 @@
-#include <system/SystemClock.h>
-
-#include <memory>
+#import "app/ConcreteAttributePath.h"
```
It's still incorrect., I'll clean these up as a touch all the files. I'm cleaning up headers as I touch files, I encourage us all to do the same.

woody-apple commented on src/darwin/Framework/Matter.xcodeproj/project.pbxproj:
```diff
@@ -109,7 +109,7 @@
 		3D3928D72BBCEA3D00CDEBB2 /* MTRAvailabilityTests.m in Sources */ = {isa = PBXBuildFile; fileRef = 3D3928D62BBCEA3D00CDEBB2 /* MTRAvailabilityTests.m */; settings = {COMPILER_FLAGS = "-UMTR_NO_AVAILABILITY -Wno-unguarded-availability-new"; }; };
 		3D4733AF2BDF1B80003DC19B /* MTRSetupPayloadTests.m in Sources */ = {isa = PBXBuildFile; fileRef = 3D4733AE2BDF1B80003DC19B /* MTRSetupPayloadTests.m */; };
-		3D4733B32BE2D1DA003DC19B /* MTRUtilities.h in Headers */ = {isa = PBXBuildFile; fileRef = 3D4733B22BE2D1CF003DC19B /* MTRUtilities.h */; };
+		3D4733B32BE2D1DA003DC19B /* MTRUtilities.h in Headers */ = {isa = PBXBuildFile; fileRef = 3D4733B22BE2D1CF003DC19B /* MTRUtilities.h */; settings = {ATTRIBUTES = (Public, ); }; };
```
I think you'll find it is:

```

/**
 * Calls `-[NSObject isEqual:]` but returns YES when comparing nil to nil.
 */
MTR_EXTERN BOOL MTREqualObjects(id<NSObject> _Nullable a, id<NSObject> _Nullable b);
```

That said, if we don't want to make it public, that's fine too.

woody-apple replied:
Will move this to _Internal for now, given we don't need to fix the examples (which have the same issue)

woody-apple commented:
> So the real fix here is making sure we don't pass nil to `isEqualToNumber:`, right? The rest is fixing other correctness issues and/or other issues altogether.

The fix here is to fix a crasher, where our comparators compare nil, which crashes.

When seeing this, I fixed this same anti-pattern in files that use it. I'm also cleaning up imports, since the files did not import the utilities function, so builds failed. I moved our includes to be imports, as is correct in the files affected. I mentioned I'll have a followup to fix all of these.

> Would be good for the commit message to make this clear....

I don't think commit messages require review, as this is a squashed commit anyway from my attempts to fix builds as I discovered import breakage (well, lack of imports).

bzbarsky-apple commented on src/darwin/Framework/Matter.xcodeproj/project.pbxproj:
Not sure why there's a diff here, but it's not actually changing anything, right?

woody-apple replied:
No idea why, but re-saving xcode did this on a newer version
